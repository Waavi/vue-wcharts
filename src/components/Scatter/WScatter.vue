<template>
    <g v-if="active">
        <WSpread
            v-if="line"
            axis="x"
            :transition="transition"
        >
            <path
                :d="linePath"
                :style="{ transition: `all 250ms ${transEffect}` }"
                :stroke="lineStylesCmp.stroke ? lineStylesCmp.stroke : fillColor"
                :stroke-width="lineStylesCmp.strokeWidth"
                :stroke-dasharray="lineStylesCmp.strokeDasharray"
                fill="none"
            />
        </WSpread>
        <g>
            <template
                v-for="dotItem in dotsData"
            >
                <slot
                    name="dot"
                    :dot="dotItem"
                    :styles="{
                        stroke: stylesCmp.stroke ? stylesCmp.stroke : fillColor,
                        fill: stylesCmp.fill ? stylesCmp.fill : fillColor,
                        strokeWidth: stylesCmp.strokeWidth,
                        opacity: stylesCmp.opacity,
                        radius: dotItem.z,
                        hoverRadius: dotItem.z,
                    }"
                    :Chart="Chart"
                    :transition="transition"
                >
                    <WDot
                        :key="`dot${dotItem.cartesianIndex}${dotItem.index}`"
                        :index="dotItem.index"
                        :cartesianIndex="dotItem.cartesianIndex"
                        :x="dotItem.x"
                        :y="dotItem.y"
                        :info="dotItem.info"
                        :styles="{
                            stroke: stylesCmp.stroke ? stylesCmp.stroke : fillColor,
                            fill: stylesCmp.fill ? stylesCmp.fill : fillColor,
                            strokeWidth: stylesCmp.strokeWidth,
                            opacity: stylesCmp.opacity,
                            radius: dotItem.z,
                            hoverRadius: dotItem.z,
                        }"
                        :transition="`all 250ms ${transEffect}`"
                        @onClick="$emit('onClickDot', $event)"
                    />
                </slot>
            </template>
        </g>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import sortBy from 'lodash.sortby'
import d3Line from 'd3-shape/src/line'
import { monotoneX as curveMonotoneX } from 'd3-shape/src/curve/monotone'
import { WDot } from '../Common'
import animationMixin from '../../mixins/animation'
import { WSpread } from '../../transitions'

import { isFunc } from '../../utils/checks'

const lineStylesDefaultProp = {
    fill: '',
    stroke: '',
    strokeWidth: 1,
    strokeDasharray: '0',
}

const stylesDefaultProp = {
    fill: '',
    stroke: '',
    strokeWidth: 0,
    radius: 8,
    opacity: 0.8,
}

export default {
    name: 'WScatter',
    type: 'cartesian',
    components: {
        WDot,
        WSpread,
    },
    mixins: [animationMixin],
    inject: ['Chart'],
    props: {
        legend: VueTypes.string,
        curve: VueTypes.oneOfType([VueTypes.bool, VueTypes.func]).def(false),
        line: VueTypes.bool.def(false),
        lineStyles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            strokeDasharray: VueTypes.string,
        }).def(() => ({
            ...lineStylesDefaultProp,
        })),
        styles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            radius: VueTypes.number,
            opacity: VueTypes.number,
        }).def(() => ({
            ...stylesDefaultProp,
        })),
    },
    computed: {
        index () {
            return this.$vnode.index
        },
        active () {
            return this.Chart.activeElements.includes(this.index)
        },
        lineStylesCmp () {
            return {
                ...lineStylesDefaultProp,
                ...this.lineStyles,
            }
        },
        stylesCmp () {
            return {
                ...stylesDefaultProp,
                ...this.styles,
            }
        },
        fillColor () {
            return this.Chart.colors[this.index]
        },
        lineData () {
            // We sort values becouse we want a left to right line
            return sortBy(this.Chart.dataset, this.Chart.axis.x.datakey).map((item, index) => ({
                x: item[this.Chart.axis.x.datakey],
                y: item[this.Chart.axis.y.datakey],
            }))
        },
        dotsData () {
            const {
                dataset, xScale, yScale, zScale, axis, colors,
            } = this.Chart
            const color = colors[this.index]
            return dataset.map((item, index) => ({
                index,
                cartesianIndex: this.index,
                x: xScale(item[axis.x.datakey]),
                y: yScale(item[axis.y.datakey]),
                z: axis.z.datakey ? zScale(item[axis.z.datakey]) : this.stylesCmp.radius,
                value: item[axis.y.datakey],
                info: {
                    id: index,
                    label: item.name || '',
                    value: [
                        this.generateAxisValue(axis.x, item[axis.x.datakey], color),
                        this.generateAxisValue(axis.y, item[axis.y.datakey], color),
                        ...(axis.z.datakey ? [this.generateAxisValue(axis.z, item[axis.z.datakey], color)] : []),
                    ],
                },
            }))
        },
        genLine () {
            return d3Line()
                .x(d => this.Chart.xScale(d.x))
                .y(d => this.Chart.yScale(d.y))
        },
        linePath () {
            if (this.curve === false) return this.genLine(this.lineData)
            const curveFn = isFunc(this.curve) ? this.curve : curveMonotoneX
            return this.line ? this.genLine.curve(curveFn)(this.lineData) : null
        },
    },
    methods: {
        generateAxisValue ({ name, datakey }, data, color) {
            return ({
                name, data, color, value: `${name}: ${data}`,
            })
        },
    },
}
</script>
