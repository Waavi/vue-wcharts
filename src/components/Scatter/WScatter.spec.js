import { shallowMount } from '@vue/test-utils'
import WScatter from './WScatter.vue'

describe('Components/WScatter', () => {
    const propsData = {
        line: true,
    }

    const provide = {
        Chart: {
            legends: ['One Bar', 'Two Bar'],
            space: [20, 20, 20, 20],
            activeElements: [0, 1],
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WScatter, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })
})
