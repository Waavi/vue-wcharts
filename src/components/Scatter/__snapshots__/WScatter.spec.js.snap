// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WScatter Should be render correctly 1`] = `
<g>
  <!---->
  <g>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="0" cartesianIndex="0" x="4000" y="2400" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2300" hoverRadius="2300" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2300" value="2400" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="1" cartesianIndex="0" x="3000" y="2210" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="1398" hoverRadius="1398" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="1398" value="2210" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="2" cartesianIndex="0" x="9800" y="2200" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="100" hoverRadius="100" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="100" value="2200" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="3" cartesianIndex="0" x="2780" y="3908" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2000" hoverRadius="2000" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2000" value="3908" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="4" cartesianIndex="0" x="1890" y="4800" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="1700" hoverRadius="1700" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="1700" value="4800" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="5" cartesianIndex="0" x="2390" y="3800" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2500" hoverRadius="2500" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2500" value="3800" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="6" cartesianIndex="0" x="3490" y="4300" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2100" hoverRadius="2100" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2100" value="4300" opacity="0.8"></WDot-stub>
  </g>
</g>
`;

exports[`Components/WScatter Shouldn be render with line 1`] = `
<g>
  <WSpread-stub axis="x" transition="all 1s ease">
    <path d="M1890,4800L2390,3800L2780,3908L3000,2210L3490,4300L4000,2400L9800,2200" fill="none" stroke="#48c0b6" strokeWidth="1" strokeDasharray="0" style="transition: all 250ms ease;"></path>
  </WSpread-stub>
  <g>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="0" cartesianIndex="0" x="4000" y="2400" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2300" hoverRadius="2300" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2300" value="2400" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="1" cartesianIndex="0" x="3000" y="2210" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="1398" hoverRadius="1398" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="1398" value="2210" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="2" cartesianIndex="0" x="9800" y="2200" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="100" hoverRadius="100" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="100" value="2200" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="3" cartesianIndex="0" x="2780" y="3908" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2000" hoverRadius="2000" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2000" value="3908" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="4" cartesianIndex="0" x="1890" y="4800" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="1700" hoverRadius="1700" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="1700" value="4800" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="5" cartesianIndex="0" x="2390" y="3800" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2500" hoverRadius="2500" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2500" value="3800" opacity="0.8"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="6" cartesianIndex="0" x="3490" y="4300" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="2100" hoverRadius="2100" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="2100" value="4300" opacity="0.8"></WDot-stub>
  </g>
</g>
`;

exports[`Components/WScatter Shouldn't be render 1`] = `undefined`;
